Docs generation for the Contracts + GH action for detecting missing Natspec comments.

indeavr opened the pull request:
Contains the dodoc output with the purpose of quick feedback. 

Description: 
1) Docs generation from Natspec comments using [dodoc harhat plugin ](https://github.com/primitivefinance/primitive-dodoc)
2) Task (at compilation) checking for public/external components with missing Natspec comments.

Resolves #1933

smartcontracts commented:
This is awesome, thank you! Since this is still a WIP, I'm going to convert this PR into a draft. Feel free to convert it back once you're ready for review.

indeavr commented:
I've just added the validation script that goes through the output and checks if there are any missing Natspec comments. 
The task is currently ran on compilation & can be configured to throw. (no custom GH action will be required)

Next:
1) Validate that the Dodoc md format is okay with Optimism.
* Feedback required.
--> If not ok -> I can edit the dodoc source (fork / PR)

2) Add Config to the validate-output task
--> error = throw or warn. 
--> checks = list of enabled checks.
--> include; exclude; runOnCompile;

3) Beautify the console output. 

4) Make a separate package (hardhat plugin) with the validate-output task. (optional)

elenadimitrova commented on packages/contracts/hardhat.config.ts:
```diff
@@ -108,6 +110,10 @@ const config: HardhatUserConfig = {
   },
+  dodoc: {
+    runOnCompile: true,
+    testMode: false,
```
Since test mode is correctly disabled here we shouldn't have the docs jsons checked in. Those are only generated for testing purposes. Please remove

elenadimitrova commented on packages/contracts/hardhat.config.ts:
```diff
@@ -108,6 +110,10 @@ const config: HardhatUserConfig = {
+  dodoc: {
+    runOnCompile: true,
+    testMode: false,
+  }
```
We should exclude the following contracts from the docs generation:
```
test-helpers
test-libraries
```

elenadimitrova commented on packages/contracts/hardhat.config.ts:
```diff
@@ -21,6 +21,8 @@ import './tasks/validate-chugsplash-dictator'
 import './tasks/withdraw-fees'
 import 'hardhat-gas-reporter'
+import '@primitivefi/hardhat-dodoc';
+import './tasks/validate-output'
```
@smartcontracts can you please review this approach for using a hardhat plugin and the script itself? This should aim to raise warnings or errors (we need to be able to run it initially in warning mode until we fix all natspec comments) on all contracts that miss natspec comments on `external` and `public` functions (as a minimum, although there might be more we could output). 
I've included this in the scope because as good as docs are they might still end up containing missing documentation unless we enforce it in build.

indeavr commented:
Update:

* Added the config options for the output-checker. 
* Added one extra check = "Compilation Warnings" in build.
* Refactored based on Elena's feedback :)

Next:
* Get Feedback on approach & output (both md and logs). 
* Move code to a npm package :)

indeavr commented:
> Update:
> 
> * Added the config options for the output-checker.
> * Added one extra check = "Compilation Warnings" in build.
> * Refactored based on Elena's feedback :)
> 
> Next:
> 
> * Get Feedback on approach & output (both md and logs).
> * Let me know if there are any other requirnments (How would the docs be deployed live ?).
> * Thorough Testing.
> * Move code to a npm package :)



> Update:
> 
> * Added the config options for the output-checker.
> * Added one extra check = "Compilation Warnings" in build.
> * Refactored based on Elena's feedback :)
> 
> Next:
> 
> * Get Feedback on approach & output (both md and logs).
> * Let me know if there are any other requirnments (How would the docs be deployed live ?).
> * Thorough Testing.
> * Move code to a npm package :)

Update: 
* Around 50-60 "missing comments" are found by the new plugin.
* Plugin is ready to be published in npm. 
* Awaiting Feedback & Additional Work :)

indeavr commented on packages/contracts/hardhat.config.ts:
```diff
@@ -108,6 +110,30 @@ const config: HardhatUserConfig = {
+      'TestLib_Buffer',
+      'TestLib_Bytes32Utils',
+      'TestLib_BytesUtils',
+      'TestLib_MerkleTree',
```
Unfortunately No. 
I'll extend Dodoc & create a PR with this feature :)

elenadimitrova replied:
That'll be nice but you don't have to do it as part of this work. We can accept current config and you can maybe reference any PR you submit to dodoc we can upgrade to later.

indeavr replied:
Great :) 
Here is the [dodoc-PR](https://github.com/primitivefinance/primitive-dodoc/pull/18).

indeavr commented on packages/contracts/hardhat.config.ts:
```diff
@@ -108,6 +110,30 @@ const config: HardhatUserConfig = {
+  },
+  outputValidator: {
+    runOnCompile: true,
+    errorMode: true,
```
Happy to hear ! I'll continue extending with anything you need. 

On your points: 
1. @inheritdoc should be working fine. I belive that the problem is with the state variables in `CrossDomainEnabled`.
2. Events = done. Public variables however are essentially functions in the abi & I can't differentiate. I'll extend the package to check in the "ast tree" for this case.
3. Should be fine now. 

* I've **disabled the check for user docs** since you're not using it anywhere (@notice). Should I return it ? 
* I've made the plugin "less-strict" accepting any Natspect comments. (config option - strict: false)

Next: 
1. @indeavr disable public vars. 
2. @elenadimitrova (after 1) Check for anything else & also review the contracts/docs folder for the actual MD output :)

indeavr replied:
Pushed ! 
The "variables: false" config option is now available. 

However be aware that there is still one issue -> "ignore state variables in base contracts". Currently only in `Lib_AddressResolver`.

I'll be doing more testing ! 
Let me know if you find any other issues :)

elenadimitrova replied:
Very nice thanks!
How difficult is it to only show a single instance of a missing natspec in base contract which is inherited elsewhere and therefore raises the error multiple times, even though only one natspec is missing. e.g. `xDomainMessageSender` is not documented in `ICrossDomainMessenger` but errors are logged from `IL1CrossDomainMessenger.sol`, `IL2CrossDomainMessenger.sol`, `L1CrossDomainMessenger.sol` and `L2CrossDomainMessenger.sol` which all inherit it.

indeavr replied:
> Very nice thanks! How difficult is it to only show a single instance of a missing natspec in base contract which is inherited elsewhere and therefore raises the error multiple times, even though only one natspec is missing. e.g. `xDomainMessageSender` is not documented in `ICrossDomainMessenger` but errors are logged from `IL1CrossDomainMessenger.sol`, `IL2CrossDomainMessenger.sol`, `L1CrossDomainMessenger.sol` and `L2CrossDomainMessenger.sol` which all inherit it.

I'll need to add some form of memoization.

elenadimitrova replied:
Don't worry if it's not straightforward. You can add it on the wish list in `hardhat-output-validator`

indeavr replied:
👍 Recursive ("inherited contracts") support for variable exclusion is now supported by the plugin
  
👎 Memoization for the output can be added on a later stage since I feel it is a lot of effort for little benefit. (prove me wrong & I'll add it now :) )

elenadimitrova replied:
What is the difference between Recursive and Memoization?

indeavr replied:
Sorry my comments were unclear ! 

You can interpret "Recursive" as `variables exclusion now works properly :D`
And the "Memoization" was meant to answer:
> How difficult is it to only show a single instance of a missing natspec in base contract ?

Answer: Quite hard.

elenadimitrova commented on packages/contracts/hardhat.config.ts:
```diff
@@ -108,6 +110,34 @@ const config: HardhatUserConfig = {
+    errorMode: true,
+    checks: {
+      events: false,
+      variables: false,
```
🎉

elenadimitrova commented on packages/contracts/hardhat.config.ts:
```diff
@@ -108,6 +110,34 @@ const config: HardhatUserConfig = {
+    ],
+  },
+  outputValidator: {
+    runOnCompile: true,
```
Also can you pls switch the `errorMode: false` to allow us to fix the natspec in a subsequent PR when we can reenable it.

elenadimitrova commented on packages/contracts/hardhat.config.ts:
```diff
@@ -108,6 +110,34 @@ const config: HardhatUserConfig = {
+    ],
+  },
+  outputValidator: {
+    runOnCompile: true,
```
Remind me how do we run this if this option is switched off in hardhat?

elenadimitrova replied:
Just saw the `docify` command, thanks :)

indeavr replied:
You can now use  - `yarn validateDocs`  👍 
be aware that Dodoc does not support this yet.

elenadimitrova commented on packages/contracts/package.json:
```diff
@@ -42,7 +42,7 @@
     "postpack": "yarn postpublish",
     "pre-commit": "lint-staged",
-    "docify": "ts-node scripts/generate-docs.ts"
+    "validateDocs": "hardhat validateOutput"
```
👍

elenadimitrova commented:
Can you please update the docs to `docs/Initializable.md` which appears to have changed. 

Another thing for the wish list is to also detect incorrect or missing function parameters or return values. I am certain this was a feature of the solidity complier before although not sure why we're not getting these reports now.

elenadimitrova commented:
There is also one failure in CI on https://github.com/ethereum-optimism/optimism/runs/4616712161?check_suite_focus=true if you can fix pls.

indeavr commented:
> Can you please update the docs to `docs/Initializable.md` which appears to have changed.
> 
> Another thing for the wish list is to also detect incorrect or missing function parameters or return values. I am certain this was a feature of the solidity complier before although not sure why we're not getting these reports now.

👍 Updated.
🥇 Added support for @return & @param  !
I as well have a memory of these errors of the compiler. However I believe the "DocstringParsingError" appears only when you misspell a @param name and not when it's missing.

elenadimitrova commented:
Nice but missing @param comments are now being inherited from external libraries, e.g. 
```
Comments Error: Function: (transferOwnership) is missing @param comment
   @ Lib_AddressManager
   --> contracts/libraries/resolver/Lib_AddressManager.sol
```

indeavr commented:
> Nice but missing @param comments are now being inherited from external libraries, e.g.
> 
> ```
> Comments Error: Function: (transferOwnership) is missing @param comment
>    @ Lib_AddressManager
>    --> contracts/libraries/resolver/Lib_AddressManager.sol
> ```

Pff this proved to be tricky. 

Should be working now :)

elenadimitrova commented:
Yep working now thanks. Checking output one by one, meanwhile I found this to be a false positive
```
Comments Error: Function: (deleteElementsAfterInclusive), param: (_globalMetadata) is missing @param comment
   @ IChainStorageContainer
   --> contracts/L1/rollup/IChainStorageContainer.sol`
```
Maybe there is a problem with overlaods?

indeavr commented:
> Yep working now thanks. Checking output one by one, meanwhile I found these to be false positives
> 
> ```
> Comments Error: Function: (deleteElementsAfterInclusive), param: (_globalMetadata) is missing @param comment
>    @ IChainStorageContainer
>    --> contracts/L1/rollup/IChainStorageContainer.sol`
> 
> Comments Error: Function: (push), param: (_globalMetadata) is missing @param comment
>    @ IChainStorageContainer
>    --> contracts/L1/rollup/IChainStorageContainer.sol
> ```
> 
> Maybe there is a problem with overloads?

Fixed. Good catch ! :) 
Dodoc has the same issue - overloads don't get their docs generated. I'll create a PR for this as well :)

elenadimitrova reviewed (approved):
Thanks for your contribution. Great work 🎉
